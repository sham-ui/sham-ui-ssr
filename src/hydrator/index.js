import Storage from './storage/service';
import ServerSideHooks from './hooks/service';
import ServerSideDOM from './dom/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Main function for setup hydrating components
 * @param {Object} DI App DI container
 */
export function setup( DI ) {
    const storage = new Storage();

    // Override default hooks service
    DI.bind(
        HOOKS_DI_KEY,

        // Pass default hooks service for proxy some methods (rehydrate & resolveID)
        new ServerSideHooks( DI.resolve( HOOKS_DI_KEY ), storage )
    );

    // Override default dom service
    DI.bind(
        DOM_DI_KEY,

        // Don't pass default dom service, because ServerSideDOM full override it
        new ServerSideDOM( storage )
    );

    return storage.root;
}


/**
 * Hydrate all components
 * @param {Object} DI App DI container
 * @return {Promise}
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import { setupHydrator, hydrate } from 'sham-ui-ssr/hydrator';
 *
 * const DI = createDI();
 * const root = setupHydrator( DI );
 * mainInitializer( DI, root );
 * start( DI );
 * hydrate( DI ).then( storage => {
 *     const { html, data } = storage.hydrate();
 * } )
 */
export function hydrate( DI ) {
    return DI.resolve( HOOKS_DI_KEY ).hydrating();
}
