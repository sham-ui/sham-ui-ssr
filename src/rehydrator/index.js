import RestoreDom from './dom/service';
import RehydrateHooks from './hooks/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Save hooks & dom services for restore after rehydrating
 * @param {Object} DI App DI container
 * @return {Function}
 */
function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

/**
 * Main function for rehydrating
 * @param {Object} DI App DI container
 * @param {Object} storage
 * @return {Function} Restore function for disable rehydrating and switch to default behavior
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import setupUnsafe from 'sham-ui-unsafe';
 * import { setupRehydrator } from 'sham-ui-ssr/rehydrator';
 *
 * const DI = createDI();
 * const data = JSON.parse( document.getElementById( 'data' ).innerText );
 * const disableRehydrating = setupRehydrator( DI, data );
 * mainInitializer( DI, root );
 * start( DI );
 * disableRehydrating();
 * setupUnsafe( DI );
 */
export function setup( DI, storage ) {
    const restore = makeRestore( DI );
    DI.bind( HOOKS_DI_KEY, new RehydrateHooks( storage ) );
    DI.bind( DOM_DI_KEY, new RestoreDom( storage ) );
    return restore;
}
