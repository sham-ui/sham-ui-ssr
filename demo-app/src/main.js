import './styles/main.scss';
import setupUnsafe from 'sham-ui-unsafe';
import { createDI, start } from 'sham-ui';
import { setup as setupHydrator, hydrate } from '../../src/hydrator/index';
import { setup as setupRehydrator } from '../../src/rehydrator/index';
import mainInitializer from './initializers/main';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

const DI = createDI();
const restoreOriginal = makeRestore( DI );

const root = setupHydrator( DI );
mainInitializer( DI, root );
start( DI );
hydrate( DI ).then(
    ( storage ) => {
        const getById = document.getElementById.bind( document );
        // eslint-disable-next-line no-console
        console.dir( storage );
        try {
            const { html, data } = storage.hydrate();
            getById( 'ssr-rendered' ).innerHTML = html;
            getById( 'raw-html' ).innerText = html;
            getById( 'data' ).innerText = JSON.stringify( JSON.parse( data ), null, 4 );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
        try {
            DI.resolve( 'sham-ui:store' ).byId.clear();
            const root = getById( 'client' );
            root.innerHTML = getById( 'ssr-rendered' ).innerHTML;
            restoreOriginal();
            const data = JSON.parse( document.getElementById( 'data' ).innerText );
            const disableRehydrating = setupRehydrator( DI, data );
            mainInitializer( DI, root );
            start( DI );
            disableRehydrating();
            setupUnsafe( DI );
            // eslint-disable-next-line no-console
            console.dir( DI.resolve( 'sham-ui:store' ).byId.keys() );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
    }
);

