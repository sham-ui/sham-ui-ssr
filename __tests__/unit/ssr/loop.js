import { compile } from 'sham-ui-test-helpers';
import { ssr } from '../../../src/testing';

beforeEach( () => {
    window.MyLi = compile`
        <li>{{ props.id + ':' + props.value }}</li>
    `;
    window.MyUl = compile`
        <ul>
            {% for element of list %}
                <MyLi props="{{ element }}"/>
            {% endfor %}
        </ul>
    `;
} );

afterEach( () => {
    delete window.MyLi;
    delete window.MyUl;
} );


it( 'should render arrays', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            <ul>
                {% for key, value of list %}
                    <li>{{ key }}:{{ value }}</li>
                {% endfor %}
            </ul>
        `,
        {
            list: [ 1, 2, 3 ]
        }
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<ul><li>0<!--0-->:<!--1-->1</li><li>1<!--0-->:<!--1-->2</li><li>2<!--0-->:<!--1-->3</li></ul>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should render arrays with externals', async() =>{
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            <div>
                {% for value of list %}
                    <p>{{ value }}{{ ext }}</p>
                {% endfor %}
            </div>
        `,
        {
            list: [ 1, 2, 3 ],
            ext: '.js'
        }
    );
    expect( meta.html ).toBe(
        '<div><p>1<!--0-->.js</p><p>2<!--0-->.js</p><p>3<!--0-->.js</p></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should iterate over objects', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            <div>
                {% for key, value of obj %}
                    {{ key }}: {{ value }};
                {% endfor %}
            </div>
        `,
        {
            obj: {
                a: 1,
                b: 2,
                c: 3
            }
        }
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<div>a<!--0-->: <!--1-->1<!--2-->; <!--3-->b<!--4-->: <!--5-->2<!--6-->; <!--7-->c<!--8-->: <!--9-->3<!--10-->; </div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should iterate over arrays without options', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            <div>
                {% for obj %}
                    {{ name }};
                {% endfor %}
            </div>
        `,
        {
            obj: [
                { name: 'a' },
                { name: 'b' },
                { name: 'c' }
            ]
        }
    );
    expect( meta.html ).toBe( '<div>a<!--0-->; <!--1-->b<!--2-->; <!--3-->c<!--4-->; </div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should delete old items from childred map with custom tag', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            <div>
                <MyUl list="{{ list }}"/>
            </div>
        `,
        {
            list: [
                {
                    id: 1,
                    value: 'a'
                },
                {
                    id: 2,
                    value: 'b'
                },
                {
                    id: 3,
                    value: 'c'
                }
            ]
        }
    );
    expect( meta.html ).toBe(
        '<div><ul><li>1:a</li><!--0--><li>2:b</li><!--0--><li>3:c</li><!--0--></ul></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should not expose local variables', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            <section>
                {% for a of as %}
                    <i>
                        {% for b of bs %}
                            <b>{{ b }}</b>
                        {% endfor %}
                    </i>
                {% endfor %}
            </section>
        `,
        {
            as: [ 'a', 'b' ],
            bs: [ 1, 2 ],
            b: 'GLOBAL'
        }
    );
    expect( meta.html ).toBe(
        '<section><i><b>1</b><b>2</b></i><i><b>1</b><b>2</b></i></section>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
