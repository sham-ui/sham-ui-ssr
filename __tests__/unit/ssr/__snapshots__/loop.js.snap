// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should delete old items from childred map with custom tag 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"O\\": {
            \\"list\\": [
                {
                    \\"id\\": 1,
                    \\"value\\": \\"a\\"
                },
                {
                    \\"id\\": 2,
                    \\"value\\": \\"b\\"
                },
                {
                    \\"id\\": 3,
                    \\"value\\": \\"c\\"
                }
            ]
        },
        \\"C\\": [
            \\"2\\",
            \\"3\\",
            \\"4\\"
        ],
        \\"N\\": [
            0
        ]
    },
    \\"2\\": {
        \\"C\\": [
            \\"5\\"
        ],
        \\"N\\": [
            1
        ]
    },
    \\"3\\": {
        \\"C\\": [
            \\"6\\"
        ],
        \\"N\\": [
            3
        ]
    },
    \\"4\\": {
        \\"C\\": [
            \\"7\\"
        ],
        \\"N\\": [
            5
        ]
    },
    \\"5\\": {
        \\"O\\": {
            \\"props\\": {
                \\"id\\": 1,
                \\"value\\": \\"a\\"
            }
        },
        \\"N\\": [
            0,
            [
                0,
                0
            ]
        ]
    },
    \\"6\\": {
        \\"O\\": {
            \\"props\\": {
                \\"id\\": 2,
                \\"value\\": \\"b\\"
            }
        },
        \\"N\\": [
            2,
            [
                2,
                0
            ]
        ]
    },
    \\"7\\": {
        \\"O\\": {
            \\"props\\": {
                \\"id\\": 3,
                \\"value\\": \\"c\\"
            }
        },
        \\"N\\": [
            4,
            [
                4,
                0
            ]
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            0
        ]
    }
}",
  "html": "
<div>
  <ul>
    <li>1:a</li>
    <!--0-->
    <li>2:b</li>
    <!--0-->
    <li>3:c</li>
    <!--0-->
  </ul>
</div>
",
}
`;

exports[`should iterate over arrays without options 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"N\\": [
            0,
            2
        ]
    },
    \\"2\\": {
        \\"N\\": [
            4,
            6
        ]
    },
    \\"3\\": {
        \\"N\\": [
            8,
            10
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\",
            \\"2\\",
            \\"3\\"
        ],
        \\"N\\": [
            0
        ]
    }
}",
  "html": "
<div>a
  <!--0-->;
  <!--1-->b
  <!--2-->;
  <!--3-->c
  <!--4-->;
</div>
",
}
`;

exports[`should iterate over objects 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"N\\": [
            0,
            2,
            4,
            6
        ]
    },
    \\"2\\": {
        \\"N\\": [
            8,
            10,
            12,
            14
        ]
    },
    \\"3\\": {
        \\"N\\": [
            16,
            18,
            20,
            22
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\",
            \\"2\\",
            \\"3\\"
        ],
        \\"N\\": [
            0
        ]
    }
}",
  "html": "
<div>a
  <!--0-->:
  <!--1-->1
  <!--2-->;
  <!--3-->b
  <!--4-->:
  <!--5-->2
  <!--6-->;
  <!--7-->c
  <!--8-->:
  <!--9-->3
  <!--10-->;
</div>
",
}
`;

exports[`should not expose local variables 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"C\\": [
            \\"3\\",
            \\"4\\"
        ],
        \\"N\\": [
            0
        ]
    },
    \\"2\\": {
        \\"C\\": [
            \\"5\\",
            \\"6\\"
        ],
        \\"N\\": [
            1
        ]
    },
    \\"3\\": {
        \\"N\\": [
            0,
            [
                0,
                0
            ]
        ]
    },
    \\"4\\": {
        \\"N\\": [
            1,
            [
                1,
                0
            ]
        ]
    },
    \\"5\\": {
        \\"N\\": [
            0,
            [
                0,
                0
            ]
        ]
    },
    \\"6\\": {
        \\"N\\": [
            1,
            [
                1,
                0
            ]
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\",
            \\"2\\"
        ],
        \\"N\\": [
            0
        ]
    }
}",
  "html": "
<section>
  <i>
    <b>1</b>
    <b>2</b>
  </i>
  <i>
    <b>1</b>
    <b>2</b>
  </i>
</section>
",
}
`;

exports[`should render arrays 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"N\\": [
            0,
            [
                0,
                0
            ],
            [
                0,
                2
            ],
            [
                0,
                4
            ]
        ]
    },
    \\"2\\": {
        \\"N\\": [
            1,
            [
                1,
                0
            ],
            [
                1,
                2
            ],
            [
                1,
                4
            ]
        ]
    },
    \\"3\\": {
        \\"N\\": [
            2,
            [
                2,
                0
            ],
            [
                2,
                2
            ],
            [
                2,
                4
            ]
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\",
            \\"2\\",
            \\"3\\"
        ],
        \\"N\\": [
            0
        ]
    }
}",
  "html": "
<ul>
  <li>0
    <!--0-->:
    <!--1-->1
  </li>
  <li>1
    <!--0-->:
    <!--1-->2
  </li>
  <li>2
    <!--0-->:
    <!--1-->3
  </li>
</ul>
",
}
`;

exports[`should render arrays with externals 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"N\\": [
            0,
            [
                0,
                0
            ],
            [
                0,
                2
            ]
        ]
    },
    \\"2\\": {
        \\"N\\": [
            1,
            [
                1,
                0
            ],
            [
                1,
                2
            ]
        ]
    },
    \\"3\\": {
        \\"N\\": [
            2,
            [
                2,
                0
            ],
            [
                2,
                2
            ]
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\",
            \\"2\\",
            \\"3\\"
        ],
        \\"N\\": [
            0
        ]
    }
}",
  "html": "
<div>
  <p>1
    <!--0-->.js
  </p>
  <p>2
    <!--0-->.js
  </p>
  <p>3
    <!--0-->.js
  </p>
</div>
",
}
`;
